### ISSUE TITLE
Add configuration to report any usage of an unawaited coroutine

### ISSUE THREAD

**Issue by maddalax** [microsoft/pyright#3563]
**Is your feature request related to a problem? Please describe.**
I've been following https://github.com/microsoft/pyright/issues/1317, but I'm hoping it would be possible to extend this to support warning/error if you have any unawaited coroutines.

<img width="1000" alt="Screen Shot 2022-06-09 at 10 50 56 AM" src="https://user-images.githubusercontent.com/11528924/172892239-0bb0c9be-cfc4-4cac-9815-5bcf6adf358c.png">

The issue being:

- the first response.json() call is being flagged as expected
- the _logger.debug(response.json()) is not flagged.

I understand we cannot make the assumption that everyone's workflow requires coroutines to be awaited / it's entirely fine to pass coroutines as a parameter, I'm more just asking if this could be something that's configurable.

**Comment by erictraut** [comment]
This would require introducing a new diagnostic rule because `reportUnusedCoroutine` would generate false positives if it included this. We do occasionally introduce new diagnostic rules, but we do so only if we have strong signal that it is generally useful. The rule you're proposing would need to be off by default, even in strict mode. That means that it wouldn't be discoverable and very few people (probably just yourself and a handful of others who happened to stumble upon it) would ever use it. So I don't think this meets our bar for adding a new diagnostic rule, at least not at this time. 

I'm going to close this for now, but we might reconsider in the future if we get signal from enough pyright users that they want this functionality.

**Comment by maddalax** [comment]
Understandable. We are converting a sync codebase to an async one, which I would imagine is somewhat of a fairly common use case as async/await becomes more mainstream in Python. But i'll wait and see if anyone else has this concern. Thanks

**Comment by nrbnlulu** [comment]
ref: https://github.com/microsoft/pyright/issues/9086#issuecomment-2376034296

> because it would lead to false positive errors in legitimate Python code

OFC it is legit though thats not your usual  usecase.
for a reference here are some linters that support this:
- dart (optional) - https://dart.dev/tools/linter-rules/unawaited_futures
- tslint (optional) - https://palantir.github.io/tslint/rules/no-floating-promises/

> If you can think of a way to make this check safer — such that it's very unlikely to produce false positive errors, then it would be easier to justify adding such a feature.

1. if you pass a coro to a function that accepts a coro thats prob not a bug
```py
async with asyncio.TaskGroup() as tg:
	sleep_for = asyncio.sleep(1)
	tg.create_task(sleep_for)
```
I'm not sure if pyright can deduce that though...

2. if you use it in a if statment
function returns `Coro | None` thats not a bug
```py
class Scrapper:
	def get_current_task(self) -> Coroutine | None:
		...
	
	async def ensure_running(self) -> None:
		if not self.get_current_task():  #  ✅
			await self.start()
```
function returns `Coro[Something]` thats prob a bug
```py
class Scrapper:
	async def get_current_task(self) -> Coroutine | None:
		...
	
	async def ensure_running(self) -> None:
		if not self.get_current_task():  #  ⚠️ you should have awaited it
			await self.start()

```
